cmd/restic/cmd_copy.go

```diff
@@ -0,0 +1,214 @@
+	f.StringVarP(&copyOptions.KeyHint, "key-hint2", "", os.Getenv("RESTIC_KEY_HINT2"), "key ID of key to try decrypting the destination repository first (default: $RESTIC_KEY_HINT2)")
+	f.StringVarP(&copyOptions.PasswordCommand, "password-command2", "", os.Getenv("RESTIC_PASSWORD_COMMAND2"), "specify a shell command to obtain a password for the destination repository (default: $RESTIC_PASSWORD_COMMAND2)")
+
+	f.StringArrayVarP(&copyOptions.Hosts, "host", "H", nil, "only consider snapshots for this `host`, when no snapshot ID is given (can be specified multiple times)")
+	f.Var(&copyOptions.Tags, "tag", "only consider snapshots which include this `taglist`, when no snapshot-ID is given")
+	f.StringArrayVar(&copyOptions.Paths, "path", nil, "only consider snapshots which include this (absolute) `path`, when no snapshot-ID is given")
+}
+
+func runCopy(opts CopyOptions, gopts GlobalOptions, args []string) error {
+	if opts.Repo == "" {
+		return errors.Fatal("Please specify a destination repository location (--repo2)")
+	}
+	var err error
+	dstGopts := gopts
+	dstGopts.Repo = opts.Repo
+	dstGopts.PasswordFile = opts.PasswordFile
+	dstGopts.PasswordCommand = opts.PasswordCommand
+	dstGopts.KeyHint = opts.KeyHint
+	if opts.password != "" {
+		dstGopts.password = opts.password
+	} else {
+		dstGopts.password, err = resolvePassword(dstGopts, "RESTIC_PASSWORD2")
+		if err != nil {
+			return err
+		}
+	}
+	dstGopts.password, err = ReadPassword(dstGopts, "enter password for destination repository: ")
+	if err != nil {
+		return err
+	}
+
+	ctx, cancel := context.WithCancel(gopts.ctx)
+	defer cancel()
+
+	srcRepo, err := OpenRepository(gopts)
+	if err != nil {
+		return err
+	}
+
+	dstRepo, err := OpenRepository(dstGopts)
+	if err != nil {
+		return err
+	}
+
+	srcLock, err := lockRepo(srcRepo)
+	defer unlockRepo(srcLock)
+	if err != nil {
+		return err
+	}
+
+	dstLock, err := lockRepo(dstRepo)
+	defer unlockRepo(dstLock)
+	if err != nil {
+		return err
+	}
+
+	debug.Log("Loading source index")
+	if err := srcRepo.LoadIndex(ctx); err != nil {
+		return err
+	}
+
+	debug.Log("Loading destination index")
+	if err := dstRepo.LoadIndex(ctx); err != nil {
+		return err
+	}
+
+	visitedTrees := restic.NewIDSet()
+	copiedSnapshots := make(map[restic.ID]*restic.ID)
+
+	for sn := range FindFilteredSnapshots(ctx, dstRepo, opts.Hosts, opts.Tags, opts.Paths, nil) {
+		if sn.Original != nil && !sn.Original.IsNull() {
+			copiedSnapshots[*sn.Original] = sn.ID()
+		}
+	}
+
+	for sn := range FindFilteredSnapshots(ctx, srcRepo, opts.Hosts, opts.Tags, opts.Paths, args) {
+		Verbosef("\nsnapshot %s of %v at %s)\n", sn.ID().Str(), sn.Paths, sn.Time)
+
+		if newID, ok := copiedSnapshots[*sn.ID()]; ok {
+			Verbosef("skipping source snapshot %s, was already copied to snapshot %s\n", sn.ID().Str(), newID.Str())
+			continue
+		}
+		if sn.Original != nil && dstRepo.Has(restic.Handle{Type: restic.SnapshotFile, Name: sn.Original.String()}) {
```

Comment:
My idea here was that snapshots IDs are random and therefore it is basically impossible for the same snapshot ID to appear in both repositories. Skipping original snapshots would only help if someone copied the raw files of a repository and then later on uses copy between both repositories. So it's probably rarely useful, but I guess it doesn't hurt. -> Is implemented.

Reply:
My intention actually was to remove this `dstRepo.Has` (which is always slow as it needs to question the backend), but I realized that this doen't work that way...

You are right that there should not be the same IDs within both repos. However, then this `if` already does not check anything: if the original snapshot exists in `dstRepo` it has not been saved with ID `sn.Original` but with a random new ID. However, if the snapshot within `dstRepo` would have set its own `sn.Original` to the original ID of this snapshot, this situation can be detected.

So instead if 
```golang
// Original does not have relevance in the new repo. Remember the source snapshot instead
		sn.Original = sn.ID()
```
below, we should use
```golang
		if sn.Originial != nil {
			sn.Original = sn.ID()
		}
```

and 
```golang
// check whether the destination has a snapshot copied from sn, whose tree is still identical
		if dstSn, ok := dstSnapshotByOriginal[*sn.ID()]; ok && dstSn.Tree.Equal(*sn.Tree) {
```
above should be changed to 
```golang
srcOriginal := *sn.ID()
if sn.Original != nil {
	srcOriginal = sn.Original
}
if dstSn, ok := dstSnapshotByOriginal[srcOriginal]; ok && dstSn.Tree.Equal(*sn.Tree) {
```

Reply:
`dstRepo.Has` for `sn.Original` handled the case when you copy snapshots from B->A and have previously copied snapshots from A->B. The main difference of always using the `Original` field is that it provides a more consistent behavior in regard to tag modifications: My previous implementation was rather inconsequential in that regard: Assume we copy a snapshot from A to B. Changing tags on the copy in B will keep `Original` and therefore the link to A. On the other hand, changing the tags on the copy in A changes the snapshot ID referenced by `Original` and therefore breaks the link.

I've added your proposed changes in a slightly expanded version: With using the `Original` field as a persistent snapshot ID, it's possible (but probably rare) for multiple snapshots to have the same ID in their `Original` field. I've therefore changed to map to store a snapshot slice.

Regarding changes to the tags of a snapshot, my feeling is that we want to copy a snapshot again once it's tags have changed. In order to avoid playing whack-a-mole with possible future PRs, there's now a `similarSnapshots` function which checks that everything except the Parent and Original fields match.

Reply:
@aawsome Can you take another look at the snapshot deduplication and check whether it now works as you'd expect?